### PR TITLE
Case search timing breadcrumbs

### PR DESCRIPTION
**Pull request by esoergel**
## Technical Summary
This PR adds timing context to the "Breadcrumbs" stored in sentry using the existing `LogLongRequestMiddleware` util, and makes some minor tweaks to that display.

This is live on staging:

image

## Safety Assurance
I'll be testing on staging

### Rollback instructions
- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

File: corehq/middleware.py [hunk 1]
```diff
@@ -120,12 +120,15 @@ def process_view(self, request, view_fn, view_args, view_kwargs):
...
         if request_timer:
-            for sub in request_timer.to_list(exclude_root=True):
-                add_breadcrumb(
-                    category="timing",
-                    message=f"{sub.name}: {sub.duration:0.3f}",
-                    level="info",
-                )
+
+            def visit(element, prefix=""):
+                name = " - ".join([prefix, element.name])
```
Comment (dannyroberts) [inline corehq/middleware.py:125]: Looks like this is the build issue:
```
    File "/mnt/commcare-hq-ro/corehq/middleware.py", line 125, in visit
      name = " - ".join([prefix, element.name])
  TypeError: sequence item 1: expected str instance, NoneType found
```

File: corehq/util/timer.py [hunk 2]
```diff
@@ -198,6 +198,26 @@ def to_list(self, exclude_root=False):
         """Get the list of ``NestableTimer`` objects in hierarchy order"""
         return self.root.to_list(exclude_root)
 
+    def add_to_sentry_breadcrumbs(self):
+        """Add timing context to the breadcrumbs section in sentry.
+
+        This must be called before sending to sentry, for instance if using
+        notify_exception rather than raising a hard error
```
Comment (esoergel) [inline corehq/util/timer.py:205]: I believe this was the problem before - the case search view called `notify_exception`, and then _afterwards_, the middleware injected this context.

I also considered making `timing_context` an optional arg to `notify_exception`, so the order dependency can be handled there, but I wasn't sure if that was a good coupling.  You could in theory use this for requests that might 500 and not use `notify_exception` at all, but I suppose that's more theoretical.  Let me know your thoughts.

**Review by orangejenny — APPROVED** [review]

**Review by orangejenny — APPROVED** [review]

File: corehq/util/timer.py (hunk 2 above)
Comment (gherceg) [reply corehq/util/timer.py:206]: I may not have the full picture but I think the theoretical scenario you highlighted sounds pretty reasonable? And therefore is worth keeping this separate from notify_exception/relying on the caller to order these correctly.

File: corehq/middleware.py [hunk 3]
```diff
@@ -336,7 +330,7 @@ def get_view_func(view_fn, view_kwargs):
         try:
             class_name = dispatcher.get_report_class_name(domain, slug)
             return to_function(class_name) if class_name else None
-        except:
+        except Exception:
```
Comment (gherceg) [inline corehq/middleware.py:333]: My ignorance to report dispatchers is pretty bad. Do we still have custom report dispatchers/is this try/except still relevant?

**Review by gherceg — APPROVED** [review]

File: corehq/middleware.py (hunk 3 above)
Comment (esoergel) [reply corehq/middleware.py:333]: Looks like it, `CustomProjectReportDispatcher` is still around.

File: corehq/tests/test_middleware.py [hunk 4]
```diff
@@ -149,7 +153,13 @@ def test_middleware_reports_slow_function_view_with_timer(self, notify_exception
         self.assertEqual(res.status_code, 200)
         notify_exception.assert_called_once()
         add_breadcrumb.assert_has_calls([
-            mock.call(category="timing", message=Regex(r"^sleep: 0.\d+"), level="info")
+            mock.call(category="timing", message=message, level="info")
+            for message in [
+                "⏱  100%  slow_function_view: 0.200s",
+                "⏱    0%   → part1: 0.000s",
+                "⏱    0%   →  → part1a: 0.000s",
+                "⏱  100%   → part2: 0.200s",
```
Comment (snopoke) [inline corehq/tests/test_middleware.py:161]: fancy 🎩

**Pull request merged**